pmd-java/src/main/java/net/sourceforge/pmd/lang/java/dfa/VariableAccessVisitor.java

```diff
@@ -76,8 +80,58 @@ private void computeNow(Node node) {
                 }
                 undefinitions.add(new VariableAccess(VariableAccess.UNDEFINITION, vnd.getImage()));
 
+                //make arraylist which could store all name occurrences
```

Comment:
We don't need so many lists... We have essentially two stages:
1. Map the name occurrences to their assignment expressions, if there are any -> `List<SimpleEntry<..,..>>`
2. Sort the name occurrences -> same List

Last step is to add the variable access in order -> we simply traverse the same list, no need to create a new one.

I've come up with the following code which only create one temporary list (occurrencesWithAssignmentExp):

``` java
                // Map the name occurrences to their assignment expressions, if any
                List<SimpleEntry<Node, NameOccurrence>> occurrencesWithAssignmentExp = new ArrayList<>();
                for (NameOccurrence occurrence : entry.getValue()) {
                    // find the nearest assignment, if any
                    Node potentialAssignment = occurrence.getLocation().getFirstParentOfAnyType(ASTStatementExpression.class,
                            ASTExpression.class);
                    while (potentialAssignment != null
                            && (potentialAssignment.jjtGetNumChildren() < 2
                                    || !(potentialAssignment.jjtGetChild(1) instanceof ASTAssignmentOperator))) {
                        potentialAssignment = potentialAssignment.getFirstParentOfAnyType(ASTStatementExpression.class,
                                ASTExpression.class);
                    }
                    // at this point, potentialAssignment is either a assignment or null
                    occurrencesWithAssignmentExp.add(new SimpleEntry<>(potentialAssignment, occurrence));
                }

                // The name occurrences are in source code order, that means, the left hand side of the
                // assignment is first. But this is not the order in which the data flows: first the
                // right hand side is evaluated before the left hand side is assigned.
                // Therefore move the name occurrences backwards if they belong to the same assignment expression.
                for (int i = 0; i < occurrencesWithAssignmentExp.size() - 1; i++) {
                    SimpleEntry<Node, NameOccurrence> oc = occurrencesWithAssignmentExp.get(i);
                    SimpleEntry<Node, NameOccurrence> nextOc = occurrencesWithAssignmentExp.get(i + 1);

                    if (oc.getKey() != null && oc.getKey().equals(nextOc.getKey())) {
                        Collections.swap(occurrencesWithAssignmentExp, i, i + 1);
                    }
                }

                for (SimpleEntry<Node, NameOccurrence> oc : occurrencesWithAssignmentExp) {
                    addAccess((JavaNameOccurrence) oc.getValue(), inode);
                }
```

I used `getFirstParentOfAnyType` to directly go to potential assignment expressions. Assignment expressions are either Expression nodes or StatementExpression node, which have the assignment operator as a second child.

Reply:
This solution extremly simplify the codes!
I've learned lots of things from this chance:)
I modified the codes according to your instruction,
and I think it looks far much better.
Thanks a lot.